warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/partition/SortedTabletLocationPartitioner.java

```diff
@@ -0,0 +1,100 @@
+        if(isAssignedPartition(tableName, index)){
+
+        }
+        assignPartitions(numPartitions, tableName, cutPointArrayLength);
+        if (isAssignedPartition(tableName, index)) {
+            return SPLIT_TO_REDUCER_MAP.get(tableName).get(index);
+        }
+        log.trace("Index was not found after recomputing the reducer map");
+        return 0;
+    }
+
+    private void assignPartitions(int numPartitions, String tableName, int cutPointArrayLength) {
+        Map<Text, String> currentTableSplitToLocation = splitToLocationMap.get().get(tableName);
+        Map<Integer,Integer> tempSplitReducerMap = new HashMap<>();
+        Text[] cutPointArray = splitsByTable.get().get(tableName);
+
+        if (cutPointArrayLength > numPartitions) {
+            mapPartitions(numPartitions, cutPointArrayLength, currentTableSplitToLocation, tempSplitReducerMap, cutPointArray);
+        } else
+            for (int i = 0; i < cutPointArrayLength; i++){
+            tempSplitReducerMap.put(i, i);
+            tempSplitReducerMap.put(-i - 1, i);
+            }
+
+            SPLIT_TO_REDUCER_MAP.put(tableName, tempSplitReducerMap);
+    }
+
+    private void mapPartitions(int numPartitions, int cutPointArrayLength, Map<Text, String> currentTableSplitToLocation, Map<Integer,Integer> tempSplitReducerMap, Text[] cutPointArray) {
+        int locationsPerSmallReducer;
+        int numReducersWithExtraLocation;
+        int firstLargeReducer;
+        int locationsAssigned = 0;
+        int assignedReducer = 0;
+        int numLocations;
+        boolean reducerFull;
+        TreeMultimap<String, Integer> locationToSplits = TreeMultimap.create();
+
+        for (int k = 0; k < cutPointArrayLength; k++) {
+            locationToSplits.put(currentTableSplitToLocation.get(cutPointArray[k]), k);
+        }
+
+        numLocations = locationToSplits.keys().size();
+        locationsPerSmallReducer = numLocations / numPartitions;
+        numReducersWithExtraLocation = numLocations % numPartitions;
+        firstLargeReducer = numPartitions - numReducersWithExtraLocation;
+
+        Iterator<String> locationIterator = locationToSplits.keySet().iterator();
+        while (locationIterator.hasNext()) {
+            Set<Integer> splitsForCurrentLocation = locationToSplits.get(locationIterator.next());
+
+            for (Integer splitIndex : splitsForCurrentLocation) {
+                tempSplitReducerMap.put(splitIndex, assignedReducer);
+                tempSplitReducerMap.put(-splitIndex - 1, assignedReducer);
+            }
+            locationsAssigned++;
+            reducerFull = ((locationsAssigned == locationsPerSmallReducer) && (assignedReducer < firstLargeReducer)) || ((locationsAssigned == locationsPerSmallReducer + 1) && (assignedReducer >= firstLargeReducer));
+            if (reducerFull)
+                assignedReducer = incrementReducer(numPartitions, assignedReducer);
```

Comment:
Can we find the "least loaded" reducer?

Reply:
Least loaded by the number of TabletLocations handled or by the number of splits per reducer?  The number of TabletLocations handled will vary by 1 at most.  I had another more complicated version of this where I made sure that the number of splits were more even per reducer by splitting them into more than one reducer per tabletlocation in case for some reason the number of tablets handled per tserver was unbalanced  Could you give me more information on what you're interested in?

Reply:
I guess a round robin distribution is sufficient if we assume that the system is balanced in the first place.